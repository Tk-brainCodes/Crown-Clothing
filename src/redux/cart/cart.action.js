import { CartActionTypes } from "./cart.action.types";

export const clearCart = () => ({
  type: CartActionTypes.CLEAR_CART,
});
