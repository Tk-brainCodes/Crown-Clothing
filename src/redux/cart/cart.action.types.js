export const CartActionTypes = {
  CLEAR_CART: "CLEAR_CART",
};
