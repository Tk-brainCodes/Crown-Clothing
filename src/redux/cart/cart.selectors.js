import {createSelector} from 'reselect';

const selectCart = state => state.cart; //main cart state

export const selectCartItems = createSelector(
    [selectCart], 
    cart => cart.cartItem //cartItem [] state
)

export const selectCartItemsCount = createSelector(
    [selectCartItems], //perform calc on the cart.cartItem[quantity]
    cartItem => cartItem.reduce((accumulatorQuantity, cartItems) => accumulatorQuantity + cartItems.quantity,0));