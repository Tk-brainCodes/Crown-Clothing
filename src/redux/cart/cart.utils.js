export const addItemToCart = (cartItem, cartItemToAdd) => {
  const existingCartItem = cartItem.find(
    cartItems => cartItems.id === cartItemToAdd.id
  )

  if (existingCartItem) {
    return cartItem.map(cartItems =>
      cartItems.id === cartItemToAdd.id
        ? { ...cartItems, quantity: cartItems.quantity + 1 }
        : cartItems
    )
  }

  return [...cartItem, { ...cartItemToAdd, quantity: 1 }]
}


export const removeItemsFromCart = (cartItem, cartItemToRemove) => {
   const existingCartItem = cartItem.find(cartItems => cartItems.id === cartItemToRemove.id);
   const cartQuantity = existingCartItem.quantity;
   if(cartQuantity === 1){
     return cartItem.filter( cartItems => cartItems.id !== cartItemToRemove.id);
   }
   return cartItem.map(cartItems => cartItems.id === cartItemToRemove.id ? {...cartItems, quantity: cartItems.quantity - 1} : cartItems )
}