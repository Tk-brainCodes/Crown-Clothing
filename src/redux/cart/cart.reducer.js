import { CartActionTypes } from "./cart.action.types";
import { addItemToCart, removeItemsFromCart } from "./cart.utils";

const INITIAL_STATE = {
  hidden: true,
  cartItem: [],
};

const cartReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case CartActionTypes.TOGGLE_CART_HIDDEN:
      return {
        ...state,
        hidden: !state.hidden,
      };
    case CartActionTypes.ADD_ITEM:
      return {
        ...state,
        cartItem: addItemToCart(state.cartItem, action.payload),
      };
    case CartActionTypes.CLEAR_ITEM_FROM_CART:
      return {
        ...state,
        cartItem: state.cartItem.filter(
          (cartItems) => cartItems.id !== action.payload.id
        ),
      };
    case CartActionTypes.REMOVE_ITEM:
      return {
        ...state,
        cartItem: removeItemsFromCart(state.cartItem, action.payload),
      };
    case CartActionTypes.CLEAR_CART:
      return {
        ...state,
        cartItem: [],
      };
    default:
      return state;
  }
};
export default cartReducer;
