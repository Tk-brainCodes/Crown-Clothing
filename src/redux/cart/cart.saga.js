import { all, call, takeLatest, put } from "redux-saga/effects";
import UserActionTypes from "../user/user.types";
import { clearCart } from "./cart.action";

export function* clearUserCart() {
  yield put(clearCart());
}

export function* onSignoutSuccess() {
  yield takeLatest(UserActionTypes.SIGN_OUT_SUCCESS, clearUserCart);
}

export function* cartySagas() {
  yield all([call(onSignoutSuccess)]);
}
