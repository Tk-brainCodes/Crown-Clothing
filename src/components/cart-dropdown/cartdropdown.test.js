import {shallow} from 'enzyme';
import React from 'react';
import { toggleCartHidden } from '../../redux/cart/cart.action';
import CartDropdown from './cart-dropdown.component';

it("expect to render CartComponent", () => {
    const CartMock = [{
        cartItem: [{
          name: 'jacket',
          price: 50,
          quantity: 1
        }],
        history: "/checkout",
        dispatch: dispatchEvent
    }]
    expect(shallow(<CartDropdown props={CartMock}/>)).toMatchSnapshot();
})