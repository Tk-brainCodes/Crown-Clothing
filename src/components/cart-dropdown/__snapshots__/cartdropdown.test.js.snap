// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render CartComponent 1`] = `ShallowWrapper {}`;
